### PR TITLE
Update adc driver to support nRF54L15

### PR DESCRIPTION
**Pull request by kl-cruz**

### CODE REVIEW

**Comment by MaureenHelm** [comment]
Needs to be rebased.

@anangl please take a look

**Comment by kl-cruz** [comment]
Twister tests issues are not related to this PR. Bug raised: https://github.com/zephyrproject-rtos/zephyr/issues/71437

**Review by nordic-segl — APPROVED** [review]

File: drivers/adc/adc_nrfx_saadc.c [hunk 1]
```diff
@@ -43,85 +61,148 @@ static struct driver_data m_data = {
...
+	uint16_t acq_time =
+		(acquisition_time == ADC_ACQ_TIME_DEFAULT
+			 ? DEFAULT_ACQ_TIME_IN_NS
+			 : (ADC_ACQ_TIME_VALUE(acquisition_time) *
+			    (ADC_ACQ_TIME_UNIT(acquisition_time) == ADC_ACQ_TIME_MICROSECONDS
+				     ? 1000
+				     : 1)));
+
+	tacq = (nrf_saadc_acqtime_t)(acq_time / MINIMUM_ACQ_TIME_IN_NS) - 1;
+	if ((tacq > SAADC_CH_CONFIG_TACQ_Max) || (acq_time < MINIMUM_ACQ_TIME_IN_NS)) {
```
Comment (anangl) [inline drivers/adc/adc_nrfx_saadc.c:107]:
```suggestion
	if ((tacq > NRF_SAADC_ACQTIME_MAX) || (acq_time < MINIMUM_ACQ_TIME_IN_NS)) {
```

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_nrfx_saadc.c [hunk 2]
```diff
@@ -168,10 +269,20 @@ static void adc_context_update_buffer_pointer(struct adc_context *ctx,
...
+		#if (NRF_SAADC_8BIT_SAMPLE_WIDTH == 8)
+		if (nrfy_saadc_resolution_get(NRF_SAADC) == NRF_SAADC_RESOLUTION_8BIT) {
+			buffer = (uint8_t *)nrfy_saadc_buffer_pointer_get(NRF_SAADC) +
+				 nrfy_saadc_amount_get(NRF_SAADC);
+		} else
+		#endif
+		{
+			buffer = (uint16_t *)nrfy_saadc_buffer_pointer_get(NRF_SAADC) +
+				 nrfy_saadc_amount_get(NRF_SAADC);
+		}
```
[the remark this answers is not in the document]
Comment (kl-cruz) [reply drivers/adc/adc_nrfx_saadc.c:283]: Looks good. It makes the code more readable

**Review by anangl — DISMISSED** [review]

**Review by nika-nordic — DISMISSED** [review]

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts [hunk 3]
```diff
@@ -147,4 +147,7 @@
 		t-enter-dpd = <10000>;
 		t-exit-dpd = <35000>;
 	};
+
+&adc {
+	status = "okay";
```
[the remark this answers is not in the document]
Comment (kl-cruz) [reply boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts:153]: Enabling it on board level is consistent with other Nordic DKs

File: drivers/adc/adc_nrfx_saadc.c [hunk 4]
```diff
@@ -16,6 +16,20 @@ LOG_MODULE_REGISTER(adc_nrfx_saadc);
...
+static const uint8_t saadc_psels[NRF_SAADC_AIN7 + 1] = {
+	[NRF_SAADC_AIN0] = NRF_PIN_PORT_TO_PIN_NUMBER(4U, 1),
+	[NRF_SAADC_AIN1] = NRF_PIN_PORT_TO_PIN_NUMBER(5U, 1),
+	[NRF_SAADC_AIN2] = NRF_PIN_PORT_TO_PIN_NUMBER(6U, 1),
+	[NRF_SAADC_AIN3] = NRF_PIN_PORT_TO_PIN_NUMBER(7U, 1),
+	[NRF_SAADC_AIN4] = NRF_PIN_PORT_TO_PIN_NUMBER(11U, 1),
+	[NRF_SAADC_AIN5] = NRF_PIN_PORT_TO_PIN_NUMBER(12U, 1),
+	[NRF_SAADC_AIN6] = NRF_PIN_PORT_TO_PIN_NUMBER(13U, 1),
+	[NRF_SAADC_AIN7] = NRF_PIN_PORT_TO_PIN_NUMBER(14U, 1),
+};
```
Comment (masz-nordic) [inline drivers/adc/adc_nrfx_saadc.c:30]: We will need to find a better solution for AIN lookup in the future, this does not scale well.

File: drivers/adc/adc_nrfx_saadc.c [hunk 5]
```diff
@@ -43,85 +61,158 @@ static struct driver_data m_data = {
...
+			    (ADC_ACQ_TIME_UNIT(acquisition_time) == ADC_ACQ_TIME_MICROSECONDS
+				     ? 1000
+				     : 1)));
+
+	tacq = (nrf_saadc_acqtime_t)(acq_time / MINIMUM_ACQ_TIME_IN_NS) - 1;
+	if ((tacq > NRF_SAADC_ACQTIME_MAX) || (acq_time < MINIMUM_ACQ_TIME_IN_NS)) {
+		result = -EINVAL;
+	} else {
+		*p_tacq_val = tacq;
+	}
```
Comment (masz-nordic) [inline drivers/adc/adc_nrfx_saadc.c:121]: I suppose we need a function for that in HAL.

File: drivers/adc/adc_nrfx_saadc.c [hunk 6]
```diff
@@ -43,85 +61,158 @@ static struct driver_data m_data = {
...
 	};
 	uint8_t channel_id = channel_cfg->channel_id;
+	uint32_t input_negative = channel_cfg->input_negative;
 
 	if (channel_id >= SAADC_CH_NUM) {
 		return -EINVAL;
 	}
 
 	switch (channel_cfg->gain) {
+#if defined(SAADC_CH_CONFIG_GAIN_Gain1_6)
```
Comment (masz-nordic) [inline drivers/adc/adc_nrfx_saadc.c:146]: MDK use should be avoided, but I don't see a real alternative for now.

**Review by masz-nordic — APPROVED** [review]
Looks good, some additions in nrfx will be needed to tidy up in the future.

**Review by anangl — APPROVED** [review]

**Pull request merged**